sklearn/tree/_tree.pyx

```diff
@@ -1274,6 +1275,192 @@ cdef class Tree:
+                                                    * nodes[left_idx].weighted_n_node_samples
+                                                )
+                                            if oob_node_values[right_idx, c, k] > 0.0 and self.value[right_value_idx] > 0.0:
+                                                importances[node.feature] += (
+                                                    (self.value[right_value_idx] * log(oob_node_values[right_idx, c, k])
+                                                    + log(self.value[right_value_idx]) * oob_node_values[right_idx, c, k])
+                                                    * nodes[right_idx].weighted_n_node_samples
+                                                )
+                                else: # Regression
+                                    importances[node.feature] += (
+                                        (node.impurity + oob_node_values[node_idx, 0, k])
+                                        * node.weighted_n_node_samples
+                                        - 
+                                        (nodes[left_idx].impurity + oob_node_values[left_idx, 0, k])
+                                        * nodes[left_idx].weighted_n_node_samples
+                                        - 
+                                        (nodes[right_idx].impurity + oob_node_values[right_idx, 0, k])
+                                        * nodes[right_idx].weighted_n_node_samples
+                                    )
+                            importances[node.feature] /= n_outputs
+                        elif method == "mdi_oob":
+                            for k in range(n_outputs):
+                                for c in range(n_classes[k]):
+                                    node_value_idx = node_idx * self.value_stride + k * max_n_classes + c
+                                    left_value_idx = left_idx * self.value_stride + k * max_n_classes + c
+                                    right_value_idx = right_idx * self.value_stride + k * max_n_classes + c
+                                    importances[node.feature] += (
+                                        (self.value[node_value_idx] - self.value[left_value_idx])
+                                        * (oob_node_values[node_idx, c , k] - oob_node_values[left_idx, c, k])
+                                        * nodes[left_idx].weighted_n_node_samples
+                                        + 
+                                        (self.value[node_value_idx] - self.value[right_value_idx]) 
+                                        * (oob_node_values[node_idx, c, k] - oob_node_values[right_idx, c, k])
+                                        * nodes[right_idx].weighted_n_node_samples
+                                    )
+                            importances[node.feature] /= n_outputs
```

Comment:
To keep the code easier to read and avoid deep indentation levels, it might be interesting to introduce inline helper functions, one for each method.